### ISSUE TITLE
Main engine crash. Tip: Restart the game

### ISSUE THREAD

**Issue by Andy3153** [bluemagic123/tModLoader#3777]
### Version

Latest 1.4.4-stable

### OS

Linux

### Platform

Steam

### This bug affects

Gameplay as a Player

### Description

I just tried playing tModLoader again for the first time after it got updated to 1.4.4 and I can't even get into the game.

### Log File

```
[18:10:30.337] [Main Thread/INFO] [tML]: Starting tModLoader client 1.4.4.9+2023.06.25.31|2023.06|stable|Stable|1cf5af1a24a961fcbe0d1d7fc31dcda160496504|5249959922336091465 built 8/12/2023 1:39AM
[18:10:30.347] [Main Thread/INFO] [tML]: Log date: 8/30/2023
[18:10:30.347] [Main Thread/INFO] [tML]: Running on Linux (v6.4.12.1) X64 NetCore 6.0.0
[18:10:30.348] [Main Thread/INFO] [tML]: FrameworkDescription: .NET 6.0.0-rtm.21522.10
[18:10:30.348] [Main Thread/INFO] [tML]: Executable: /home/andy3153/downs/tModLoader/tModLoader.dll
[18:10:30.348] [Main Thread/INFO] [tML]: Working Directory: /home/andy3153/downs/tModLoader
[18:10:30.606] [Main Thread/INFO] [tML]: Saves Are Located At: /home/andy3153h/.local/share/Terraria/tModLoader
[18:10:30.611] [Main Thread/INFO] [tML]: Display DPI: Diagonal DPI is 0. Vertical DPI is 0. Horizontal DPI is 0
[18:10:30.612] [Main Thread/DEBUG] [FNA]: Querying linked library versions...
[18:10:30.613] [Main Thread/DEBUG] [FNA]: SDL v2.24.0
[18:10:30.613] [Main Thread/DEBUG] [tML]: Native Resolve: FNA, Version=22.7.0.0, Culture=neutral, PublicKeyToken=null -> FNA3D
[18:10:30.613] [Main Thread/DEBUG] [tML]: 	attempting load /home/andy3153/downs/tModLoader/Libraries/Native/Linux/libFNA3D.so.0
[18:10:30.613] [Main Thread/DEBUG] [tML]: 	success
[18:10:30.613] [Main Thread/DEBUG] [FNA]: FNA3D v22.8.0
[18:10:30.613] [Main Thread/DEBUG] [tML]: Native Resolve: FNA, Version=22.7.0.0, Culture=neutral, PublicKeyToken=null -> FAudio
[18:10:30.613] [Main Thread/DEBUG] [tML]: 	attempting load /home/andy3153/downs/tModLoader/Libraries/Native/Linux/libFAudio.so.0
[18:10:30.614] [Main Thread/DEBUG] [tML]: 	success
[18:10:30.614] [Main Thread/DEBUG] [FNA]: FAudio v22.8.0
[18:10:30.818] [Main Thread/INFO] [tML]: Distribution Platform: Steam. Detection method: Terraria.exe or Terraria_v1.4.4.9.exe not found nearby
[18:10:30.824] [Main Thread/DEBUG] [TerrariaSteamClient]: Disabled. Launched outside steam client.
[18:10:30.836] [Main Thread/DEBUG] [tML]: ILHook Ionic.Zip.ZipEntry::ValidateOutput(string, Stream, out string) added by tModLoader
[18:10:30.889] [Main Thread/WARN] [tML]: Silently Caught Exception: 
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Xna.Framework.SDL2_FNAPlatform.ProgramInit(LaunchParameters args)
   at Microsoft.Xna.Framework.FNAPlatform..cctor()
   at Microsoft.Xna.Framework.Game..ctor(Boolean headlessMode)
   at Terraria.Main..ctor()
   at Terraria.Program.RunGame()
   at Terraria.Program.LaunchGame_(Boolean isServer)
   at Terraria.Program.LaunchGame(String[] args, Boolean monoArgs)
   at Terraria.MonoLaunch.Main_End(String[] args)
   at Terraria.MonoLaunch.<>c__DisplayClass1_0.<Main>b__0()
   at Terraria.MonoLaunch.Main(String[] args)

[18:10:30.889] [Main Thread/WARN] [tML]: Silently Caught Exception: 
System.TypeInitializationException: The type initializer for 'Microsoft.Xna.Framework.FNAPlatform' threw an exception.
   at Microsoft.Xna.Framework.Game..ctor(Boolean headlessMode)
   at Terraria.Main..ctor()
   at Terraria.Program.RunGame()
   at Terraria.Program.LaunchGame_(Boolean isServer)
   at Terraria.Program.LaunchGame(String[] args, Boolean monoArgs)
   at Terraria.MonoLaunch.Main_End(String[] args)
   at Terraria.MonoLaunch.<>c__DisplayClass1_0.<Main>b__0()
   at Terraria.MonoLaunch.Main(String[] args)

[18:10:30.890] [Main Thread/FATAL] [tML]: Main engine crash

Tip: Restart the game. If this issue persists, try uninstalling Terraria completely, then reinstalling through Steam, and then reinstalling tModLoader

System.TypeInitializationException: The type initializer for 'Microsoft.Xna.Framework.FNAPlatform' threw an exception.
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Xna.Framework.SDL2_FNAPlatform.ProgramInit(LaunchParameters args) in D:\a\tModLoader\tModLoader\FNA\src\FNAPlatform\SDL2_FNAPlatform.cs:line 223
   at Microsoft.Xna.Framework.FNAPlatform..cctor() in D:\a\tModLoader\tModLoader\FNA\src\FNAPlatform\FNAPlatform.cs:line 150
   --- End of inner exception stack trace ---
   at Microsoft.Xna.Framework.Game..ctor(Boolean headlessMode) in D:\a\tModLoader\tModLoader\FNA\src\Game.cs:line 264
   at Terraria.Main..ctor() in tModLoader\Terraria\Main.cs:line 5436
   at Terraria.Program.RunGame() in tModLoader\Terraria\Program.cs:line 257
```

### Steps to reproduce

1. Try launching the game

### Expected Behavior

Game launches

### Actual Behavior

Game crashes

### Reproduction frequency

Every single time I hit the play button

### Additional Information

I also tried downloading the latest release from GitHub, but I get the same result.

Also, initially it complained about moving the tModLoader folder with the save files, so I helped it by doing it myself. Then, the crashes started happening. Since then, I tried the advice I get from it (to reinstall both Terraria and tModLoader) but I also tried deleting the Terraria folder (`~/.local/share/Terraria`) entirely and I get the same result.

And, yes, there's a similar issue on here that's also coming from a Linux user (#3766), but looking at their `client.log`, it's something different.

**Comment by JavidPack** [comment]
Can you post the other log files too? Can you also confirm that 1.4.3 works? (Switch to the 1.4.3-legacy branch)

Also, is there a reason you aren't launching from Steam? It might help.

**Comment by Andy3153** [comment]
> Can you also confirm that 1.4.3 works?

It does not. Tried the version from Steam.

> Also, is there a reason you aren't launching from Steam? It might help.

I did that in an attempt to find the issue. First thing I tried is to launch from Steam.

Also, in an attempt to find the issue, I tried launching it with Proton, and while it did work, textures were buggy and the server didn't start.

Then, I tried switching from Hyprland (a Wayland compositor), to KDE Plasma 5, under X11. And I can say it works there. It could be a Hyprland/wlroots/Wayland in general issue?

> Can you post the other log files too?

Right now I'm trying to see if everything works under X11, so I'll replicate the bug to get the logs again later

**Comment by Andy3153** [comment]
Everything works seemingly well under X11. I don't understand much of the logs, can you tell me if they say the problem is related to displaying, creating a window, graphics etc. so that I can blame Hyprland? Then I can likely narrow the issue down further, like for example trying to run tModLoader under Sway, a different Wayland compositor, but still based on wlroots. If the problem persists further I'll try some different Wayland compositor to see if it's Wayland overall to blame.

**Comment by JavidPack** [comment]
I'm not familiar with those things, unfortunately. The error seems to suggest that the game was unable to create the game window, so it could be related to your windowing system. I don't know the specifics of why it isn't able to do that. Does Terraria itself work? tModLoader and Terraria on linux should be fairly similar in compatibility in terms of graphics. 

A note about Proton, it usually results in errors and if you used proton to launch the game you'll need to delete the dotnet folder in the install folder after disabling it.

**Comment by Andy3153** [comment]
> Does Terraria itself work?

Classic Terraria does work indeed, regardless of windowing system.

> A note about Proton, it usually results in errors and if you used proton to launch the game you'll need to delete the dotnet folder in the install folder after disabling it.

You don't need to worry about that, I deleted the entire save folder every time I had a new attempt at launching the game.

> The error seems to suggest that the game was unable to create the game window, so it could be related to your windowing system.

Great, thanks, I have something to test, then. I'll get to testing later, I'm playing with my friends right now lol.

For anyone keeping track/trying to replicate/help with testing/whatever, I'll firstly test Sway, the largest wlroots-based Wayland compositor to see if it's Hyprland or wlroots, then I'll probably go with other compositors to see if it's wlroots or the entirety of Wayland.

Something to note is that I remember checking if the window reporting the crash happens is a native Wayland window, or running through XWayland, a way to run X11 programs inside Wayland, and from what I remember it was running through XWayland, so it might also be causing problems.

**Comment by Andy3153** [comment]
Okay so... I found the (temporary) solution.

I was reading through my own crash logs I posted earlier and I saw `SDL` in there, and that rang a bell.

Back when I set my Hyprland config the first time, I [followed the wiki](https://wiki.hyprland.org/Configuring/Environment-variables/#toolkit-backend-variables) and I set an environment variable that made SDL try to create all the windows as native Wayland windows (`SDL_VIDEODRIVER=wayland`). That doesn't work, and it creates problems, if the game uses an old SDL2 version, as the wiki actually says.

The solution is to either not set that anymore, or, go to any game that has problems and set the environment variable to `SDL_VIDEODRIVER=x11` prior to launching the game (like, for example, in Steam, inside the launch options), and now it works for me.

The real solution theoretically would be for tModLoader to update to the latest SDL library, but this is a good enough solution until the time comes. I imagine it isn't of very high priority to do this, but it's something to be considered.

**Comment by Chicken-Bones** [comment]
@JavidPack we can easily detect this environment variable and replace it (or log it as a warning). Good reminder to post `environment.log` as well so we can check SDL related variables

**Comment by Andy3153** [comment]
Personally I'd recommend against overriding the variable without at least notifying the user. I would make tmodloader say the user should check that environment variable in the conditions I got my crash/whenever SDL crashes on Linux with Wayland/whenever Wayland is detected and a crash occurs.

> Good reminder to post environment.log as well so we can check SDL related variables

I probably should've posted every log file, but I went with the bug report template you have on here.
